cpp/server/server.h

```diff
@@ -0,0 +1,346 @@
+  if (FLAGS_watchdog_timeout_is_fatal) {
+    LOG(FATAL) << "Watchdog timed out, killing process.";
+  } else {
+    LOG(INFO) << "Watchdog timeout out, ignoring.";
+  }
+}
+
+
+void locking_function(int mode, int n, const char* file, int line) {
+  static std::mutex* openssl_locks = new std::mutex[CRYPTO_num_locks()];
+  if (mode & CRYPTO_LOCK) {
+    openssl_locks[n].lock();
+  } else {
+    openssl_locks[n].unlock();
+  }
+}
+
+
+void threadid_function(CRYPTO_THREADID* id) {
+#ifdef PTHREAD_T_IS_POINTER
+  CRYPTO_THREADID_set_pointer(id, pthread_self());
+#else
+  CRYPTO_THREADID_set_numeric(id, pthread_self());
+#endif
+}
+
+
+template <class Logged>
+std::string GetNodeId(Database<Logged>* db) {
+  std::string node_id;
+  if (db->NodeId(&node_id) != Database<LoggedCertificate>::LOOKUP_OK) {
+    node_id = cert_trans::UUID4();
+    LOG(INFO) << "Initializing Node DB with UUID: " << node_id;
+    db->InitializeNode(node_id);
+  } else {
+    LOG(INFO) << "Found DB with Node UUID: " << node_id;
+  }
+  return node_id;
+}
+
+
+}  // namespace
+
+
+// static
+template <class Logged>
+void Server<Logged>::StaticInit() {
+  // Set-up OpenSSL for multithreaded use:
+  CRYPTO_THREADID_set_callback(threadid_function);
+  CRYPTO_set_locking_callback(locking_function);
+
+  OpenSSL_add_all_algorithms();
+  ERR_load_crypto_strings();
+  cert_trans::LoadCtExtensions();
+  evthread_use_pthreads();
+
+  CHECK_NE(SIG_ERR, std::signal(SIGALRM, &WatchdogTimeout));
+}
+
+
+template <class Logged>
+Server<Logged>::Server(const Options& opts,
+                       std::shared_ptr<libevent::Base> event_base,
+                       Database<Logged>* db, EtcdClient* etcd_client,
+                       UrlFetcher* url_fetcher, LogSigner* log_signer,
+                       CertChecker* cert_checker)
+    : options_(opts),
+      event_base_(event_base),
+      event_pump_(new libevent::EventPumpThread(event_base_)),
+      http_server_(*event_base_),
+      db_(CHECK_NOTNULL(db)),
+      cert_checker_(cert_checker),
+      node_id_(GetNodeId(db_)),
+      url_fetcher_(url_fetcher),
+      etcd_client_(etcd_client),
+      election_(event_base_, etcd_client_, options_.etcd_root + "/election",
+                node_id_),
+      internal_pool_(8),
+      server_task_(&internal_pool_),
+      consistent_store_(&election_,
+                        new EtcdConsistentStore<LoggedCertificate>(
+                            event_base_.get(), &internal_pool_, etcd_client_,
+                            &election_, options_.etcd_root, node_id_)),
+      frontend_((log_signer && cert_checker)
+                    ? new Frontend(new CertSubmissionHandler(cert_checker),
+                                   new FrontendSigner(db_, &consistent_store_,
+                                                      log_signer))
+                    : nullptr),
+      http_pool_(options_.num_http_server_threads),
+      json_output_(event_base_.get()) {
+  CHECK_LT(0, options_.port);
```

Comment:
Could do, but we'd have already exploded by then if it were null because of the deref at 217...